docs/spec_draft.md

```diff
@@ -341,6 +351,174 @@ only exists to establish data dependencies from `result` to `inputs`.
+     follows:
+     ```python
+     groups = []
+     for replica_group in replica_groups:
+       sub_group = []
+       for global_device_id in replica_group:
+         sub_group.append(ei(global_device_id // num_pids, global_device_id % num_pids))
+       groups.append(sub_group)
+     return groups
+     ```
+    * For example, assuming `num_pids = 2` and `replica_groups = [[4, 5], [6, 7]]`:
+      `groups` formed: `[ei(2, 0), ei(2, 1)] and [ei(3, 0), ei(3, 1)]`.
+
+For each group `G` $\in$ `groups`, the `result`, to be visible by each member of
+`G`, is given by:
+`concatenate(operands, all_gather_dim)`, where `operands` =
+{ `operand` corresponding to the execution instances, following their order, in
+`G` }.
+
+### Inputs
+
+| Name                    | Type                                                                                                                                         |
+|-------------------------|----------------------------------------------------------------------------------------------------------------------------------------------|
+| `operand`               | tensor of any supported type                                                                                                                 |
+| `all_gather_dim`        | constant of type `si64`                                                                                                                      |
+| `replica_groups`        | 2-dimensional tensor constant of type `si64`                                                                                                  |
+| `channel_handle`        | constant of type struct { `channel-id`: `si64`, `channel-type`: enum of `CHANNEL_TYPE_INVALID`, `DEVICE_TO_DEVICE`, `DEVICE_TO_HOST`, `HOST_TO_DEVICE` } |
+| `use_global_device_ids` | constant of type `boolean`                                                                                                                   |
+
+### Outputs
+
+| Name     | Type                         |
+|----------|------------------------------|
+| `result` | tensor of any supported type |
+
+### Constraints
+
+  * (C1) `all_gather_dim` $\in$ [0, rank(`operand`)).
+  * (C2) dim(`operand`, `all_gather_dim`) $\ne$ 0.
+  * (C3) On `replica_groups`:
+    * All the subgroups in `replica_groups` have the same size.
+    * The replica ids contains all values in the range [0, N) where N is the
+      number of replicas in `replica_groups`.
+    * An empty `replica groups` is allowed only when `use_global_device_ids` is
+      `false`.
+  * (C4) dim(`result`, `all_gather_dim`) % dim(`operand`, `all_gather_dim`) $=$ 0.
+  * (C5) element_type(`result`) $=$ element_type(`operand`).
+  * (C6) rank(`result`) $=$ rank(`operand`).
+  * (C7) dim(`result`, i) is given by:
+    * dim(`operand`, i), i $\ne$ `all_gather_dim`.
+    * `shard_count` * dim(`operand`, i), otherwise, where `shard_count` $=$
+      dim(`result`, `all_gather_dim`) / dim(`operand`, `all_gather_dim`).
```

Comment:
From https://github.com/tensorflow/tensorflow/blob/72d57d68121257c089a39cf99c406591ad9ccb5d/tensorflow/compiler/xla/service/shape_inference.cc#L2091

we have `dim(result, all_gather_dim) = shard_count * dim(operand, all_gather_dim)`.

From the 
[definition of subgroup_size](https://github.com/tensorflow/tensorflow/blob/20c6943d3cd7e07da162f7778a0af5d3776274b4/tensorflow/compiler/xla/service/hlo_verifier.cc#L432) and the fact that  [subgroup_size == shard_count](https://github.com/tensorflow/tensorflow/blob/20c6943d3cd7e07da162f7778a0af5d3776274b4/tensorflow/compiler/xla/service/hlo_verifier.cc#L436), we can say that

`dim(result, all_gather_dim)` is given by
  * `dim(replica_groups, 1) * number_partitions * dim(operand, all_gather_dim)`, for process groups formation strategy $=$  `cross_replica_and_partition,
  * `dim(replica_groups, 1) * dim(operand, all_gather_dim)`, otherwise.

Reply:
Good catch - my initial suggestion was wrong! Let's try to compress this perhaps to avoid duplication with e.g. AllToAllOp? For example, how about `dim(operand, all_gather_dim) * dim(process_groups, 1)`?